awslogs driver did not create logging group

didzis opened the pull request:
Currently using awslogs driver with non-existing logging group leads to failing to start a container. Added code that first creates logging group.

Amazon's own awslogs agent is able to create logging group if it does not exist.

Signed-off-by: Didzis Gosko didzis.gosko@gmail.com

didzis committed:
awslogs driver: will try to create AWS logging group first instead of just failing to run container if logging group does not exist

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

didzis committed:
Reformatted with gofmt -s -w

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

didzis committed:
Added unit tests

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

didzis committed:
Temporary workaround for unit-test hanging on awslogs: return dummy nils in CreateLogGroup.

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

thaJeztah commented:
ping @samuelkarp can you have a look at this one?

samuelkarp commented:
@didzis Thanks for taking a look at this and @thaJeztah thanks for tagging me!

For some background: I originally decided against this behavior due to the service limit of [500 log groups](http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/cloudwatch_limits.html) and to protect against exhausting that limit accidentally by making typos.  However, I'd be happy to defer to the Docker maintainers if this behavior is ultimately desirable.

samuelkarp commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -184,6 +185,38 @@ func (l *logStream) Close() error {
 func (l *logStream) create() error {
+
+	// Create log group first
+	{
```
Is there any particular reason why you've enclosed the added code in a new block level?

didzis replied:
@samuelkarp I enclosed the code in a new block level because I wanted to reuse the code below (that creates streams) without changing variable names - I'm not a Go developer, this was the easiest and errorless way for me.
I had a feeling that Docker's awslogs driver doesn't create groups intentionally, but I needed this feature for a real use case, where I generate group names dynamically, and not expecting as much as 500 groups to be generated. Anyway, Amazon's own awslogs agent does exactly that - creates new log group if it doesn't exist. I think it is better for a container to fail to run when group limit is reached rather than fail to run if the group doesn't exist. Or at least let the user decide what is better for him by defining some extra log driver option.

samuelkarp replied:
I think you can take it out of the block with a few minor tweaks.  Specifically:
* Some of the `:=`s in the original code (not what you added) will probably need to change to `=` as Go makes a distinction between declaration and assignment.
* You'll probably want to change both instances of `input` to have different names that are more specific to what they're doing, since a single variable cannot be multiple different types.

As I said in my first comment, I'm not too opposed to this behavior change if it solves a real problem (and it sounds like it does).  One way to change it perhaps is to enable the behavior only when a log option enables it (something like `awslogs-enable-creation=true`).

calavera replied:
can you extract this block into a function? I agree that having an option would be convenient. That way we don't change the default behavior.

samuelkarp commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -184,6 +185,38 @@ func (l *logStream) Close() error {
+			LogGroupName: aws.String(l.logGroupName),
+		}
+
+		_, err := l.client.CreateLogGroup(input)
```
I'm worried about the latency cost of an additional round-trip for every container start.  I think better behavior would be attempting `CreateLogStream` first, and falling back to `CreateLogGroup` only when the log group does not already exist.

samuelkarp commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -184,6 +185,38 @@ func (l *logStream) Close() error {
+					"logGroupName":  l.logGroupName,
+					"logStreamName": l.logStreamName,
+				}
+				if awsErr.Code() == resourceAlreadyExistsCode {
```
You'll also want to explicitly handle permission problems here, in case the IAM policy is such that `logs:CreateLogGroup` is not allowed but `logs:CreateLogStream` and `logs:PutLogEvents` are.

samuelkarp commented on daemon/logger/awslogs/cwlogsiface_mock_test.go:
```diff
@@ -49,6 +60,13 @@ func (m *mockcwlogsclient) PutLogEvents(input *cloudwatchlogs.PutLogEventsInput)
 }
 
+func (m *mockcwlogsclient) CreateLogGroup(input *cloudwatchlogs.CreateLogGroupInput) (*cloudwatchlogs.CreateLogGroupOutput, error) {
+	// m.createLogGroupArgument <- input
```
Rather than commenting out the implementation of the mock, it'd be better to adjust the expectations set in the existing unit tests to handle the new calls to `CreateLogGroup`.

didzis replied:
@samuelkarp I agree on all suggestions. Didn't know where to adjust the unit test expectations, so I just commented out for now or else the unit test hangs on this.

samuelkarp replied:
You can see an example here: https://github.com/docker/docker/blob/7a016578dbb5c9b8a43b43e1ce96785fa26baf56/daemon/logger/awslogs/cloudwatchlogs_test.go#L79-L85

This sets an expectation by way of giving the mock a result that it can return when it gets called.  You can then [read the argument(s)](https://github.com/docker/docker/blob/7a016578dbb5c9b8a43b43e1ce96785fa26baf56/daemon/logger/awslogs/cloudwatchlogs_test.go#L92-L95) and perform assertions.  It's somewhat crude since I hand-wrote it (I didn't see an existing mocking tool being used in Docker), but it seems to work.

thaJeztah commented:
Thanks for looking, @samuelkarp, I'll try to get some maintainers on this to discuss the design; decide if the 500 service limit is expected to be an issue.

didzis committed:
awslogs driver:
* added awslogs-create-group option
* updated unit tests
* updated documentation

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

cpuguy83 commented:
I think i'd rather keep the error in this case.
Since logs are generated for every container on the system, and potentially we end up calling CreateGroup every time, it seems redundant and potentially surprising.

didzis committed:
awslogs log driver: when awslogs-create-group is set to true and create
stream fails because of non-existing group, will create group, and then
try to create stream again

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23630/console):

~~~console

---


---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
Congratulations!  All Go source files are properly formatted.

---> Making bundle: validate-lint (in bundles/1.11.0-dev/validate-lint)
Errors from golint:
daemon/logger/awslogs/cloudwatchlogs.go:212:11: if block ends with a return statement, so drop this else and outdent its block

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson8094501994455525506.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/834/console):

~~~console

---
aa0163.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Wed Feb 10 16:07:30 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 16:07:31 CUT 2016 (1m 13s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson3185948321975667320.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

didzis committed:
awslogs: Fixed style.

Signed-off-by: Didzis Gosko <didzis.gosko@gmail.com>

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21602/console):

~~~console

---
aking bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/docker
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: fork/exec /usr/bin/gcc: cannot allocate memory

---
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson6324824295749176313.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/835/console):

~~~console

---
f08370.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Wed Feb 10 16:13:42 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 16:13:43 CUT 2016 (1m 19s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson8622076729448874870.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14839/console):

~~~console

---


---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
Congratulations!  All Go source files are properly formatted.

---> Making bundle: validate-lint (in bundles/1.11.0-dev/validate-lint)
Errors from golint:
daemon/logger/awslogs/cloudwatchlogs.go:212:11: if block ends with a return statement, so drop this else and outdent its block

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson1676984019483790657.sh
+ set +x


---

~~~

thaJeztah commented:
Thanks for contributing @didzis, I discussed this with some maintainers, and we prefer to keep this option explicit instead of automatically creating the groups.

I'm closing this PR, but feel free to comment if you disagree :heart:

didzis commented:
@thaJeztah One of my latest commits introduces an extra option `awslogs-create-group`, which must be explicitly set to true for a log group to be created and even then only after create stream has filed because of missing log group - there will be no extra overhead during normal container start when the log group is in place.

Please consider, that without the ability to create a log group, the behavior of awslogs log driver is inconsistent with Amazon's awslogs agent.
